File: lib/billing/standard-trial-checkout.ts
/**
 * Standard (lite) monthly trial Checkout flags.
 *
 * Default: card required (historic Stripe Checkout with payment_method_types=["card"]).
 * No-card campaign: payment_method_collection=if_required and cancel at trial end
 * if no payment method was collected.
 *
 * Card-required Standard monthly trial: 3 days.
 * No-card campaigns (/landing/grandmother-free, /landing/origins-free, hero-free landings): 7 days.
 */

export type CheckoutPaymentMode = "card_required" | "no_card_trial"

export const DEFAULT_STANDARD_TRIAL_DAYS = 3
export const DEFAULT_NO_CARD_TRIAL_DAYS = 7

function parsePositiveInt(raw: string | undefined, fallback: number): number {
  const parsed = parseInt(raw || "", 10)
  if (!Number.isFinite(parsed) || parsed < 0) return fallback
  return parsed
}

export function resolveStandardTrialDays(noCard: boolean): number {
  return noCard
    ? parsePositiveInt(process.env.STANDARD_NO_CARD_TRIAL_DAYS, DEFAULT_NO_CARD_TRIAL_DAYS)
    : parsePositiveInt(process.env.STANDARD_TRIAL_DAYS, DEFAULT_STANDARD_TRIAL_DAYS)
}

export function isStandardMonthlyTrial(
  tier: string | undefined,
  billingCycle: string | undefined,
  trialDays: number
): boolean {
  const t = (tier || "").toLowerCase()
  const cycle = (billingCycle || "").toLowerCase()
  return (
    trialDays > 0 &&
    cycle === "monthly" &&
    (t === "lite" || t === "standard" || t === "basic" || t === "starter")
  )
}

export function resolveCheckoutPaymentMode(opts: {
  tier: string
  billingCycle: string
  trialDays: number
  trialWithoutCardRequested: boolean
}): CheckoutPaymentMode {
  if (
    opts.trialWithoutCardRequested &&
    isStandardMonthlyTrial(opts.tier, opts.billingCycle, opts.trialDays)
  ) {
    return "no_card_trial"
  }
  return "card_required"
}

export function stripeTrialSettingsForNoCard(): {
  end_behavior: { missing_payment_method: "cancel" }
} {
  return {
    end_behavior: { missing_payment_method: "cancel" },
  }
}

export function campaignCancelUrl(siteUrl: string, cancelPath: unknown): string | null {
  if (typeof cancelPath !== "string") return null
  if (!cancelPath.startsWith("/") || cancelPath.startsWith("//")) return null
  if (cancelPath.includes("://") || cancelPath.includes("\\")) return null
  const origin = siteUrl.replace(/\/$/, "")
  return `${origin}${cancelPath}`
}

File: llmhive/src/llmhive/app/billing/standard_trial_checkout.py
"""Standard (lite) monthly trial Checkout Session flags.

Card-required trial is the historic default (``payment_method_types=["card"]``).
No-card trial uses Stripe ``payment_method_collection=if_required`` plus
``trial_settings.end_behavior.missing_payment_method=cancel`` so access ends
after 3 days unless the customer adds a payment method.
"""
from __future__ import annotations

from typing import Any, Dict, Mapping, MutableMapping


def is_standard_monthly_trial(tier: str, billing_cycle: str) -> bool:
    return (tier or "").strip().lower() in {"lite", "standard", "basic", "starter"} and (
        billing_cycle or ""
    ).strip().lower() == "monthly"


def apply_trial_subscription_data(
    subscription_data: MutableMapping[str, Any],
    *,
    trial_days: int,
    no_card: bool,
) -> MutableMapping[str, Any]:
    if trial_days <= 0:
        return subscription_data
    subscription_data["trial_period_days"] = trial_days
    metadata = dict(subscription_data.get("metadata") or {})
    metadata["is_trial"] = "true"
    if no_card:
        metadata["trial_without_card"] = "true"
        subscription_data["trial_settings"] = {
            "end_behavior": {"missing_payment_method": "cancel"}
        }
    subscription_data["metadata"] = metadata
    return subscription_data


def apply_session_payment_collection(
    session_kwargs: MutableMapping[str, Any],
    *,
    no_card: bool,
) -> MutableMapping[str, Any]:
    """Mutate Checkout Session create kwargs. Card-required path stays explicit."""
    if no_card:
        session_kwargs.pop("payment_method_types", None)
        session_kwargs["payment_method_collection"] = "if_required"
    else:
        session_kwargs["payment_method_types"] = ["card"]
        session_kwargs.pop("payment_method_collection", None)
    return session_kwargs


def campaign_cancel_url(success_url: str, cancel_path: str | None) -> str | None:
    """Allow only same-origin relative paths such as ``/landing/grandmother-free``."""
    if not cancel_path or not cancel_path.startswith("/") or cancel_path.startswith("//"):
        return None
    if "://" in cancel_path or "\\" in cancel_path:
        return None
    from urllib.parse import urlparse

    parsed = urlparse(success_url or "")
    if not parsed.scheme or not parsed.netloc:
        return None
    return f"{parsed.scheme}://{parsed.netloc}{cancel_path}"


def customer_used_standard_trial(subscriptions: Mapping[str, Any] | list) -> bool:
    """True if Stripe already recorded a Standard trial for this customer."""
    rows = subscriptions
    if isinstance(subscriptions, dict):
        rows = subscriptions.get("data") or subscriptions.get("subscriptions") or []
    if not isinstance(rows, list):
        return False
    for sub in rows:
        if not isinstance(sub, dict):
            continue
        metadata = sub.get("metadata") or {}
        if str(metadata.get("is_trial") or "").lower() == "true":
            return True
        if str(metadata.get("trial_without_card") or "").lower() == "true":
            return True
        if sub.get("trial_start") or sub.get("trial_end"):
            tier = str(metadata.get("tier") or "").lower()
            if tier in {"lite", "standard", "basic", "starter", ""}:
                return True
    return False
